### DISCUSSION TITLE
Using PushManager in CodeBehind for Firebase

### DISCUSSION THREAD

**Discussion by konrad79** [shinyorg/shiny#1491] [Q&A]
I want to use Firebase Notifications in my Blazor App.
In my MauiProgram I register the firebase messaging with:
```
var config = new Shiny.Push.FirebaseConfiguration(
    false,
    AndroidAppId,
    ProjectNumber,
    ProjectId,
    ApiKey
);
builder.Services.AddShinyService<MyPushDelegate>();
builder.Services.AddPushFirebaseMessaging(config);
```

And since I want to use the PushManager to register my app to firebase, I add this for dependency injection as well in MauiProgram:

`builder.Services.AddSingleton<Shiny.Push.IPushManager, Shiny.Push.PushManager>();`

And in the code behind of my Razor page, I injected the PushManager with:

```
public class HomeBase: ComponentBase
{
...
    [Inject]
    public Shiny.Push.IPushManager? MyPushManager { get; set; }

    public void Register()
    {
        var result = PushManager?.RequestAccess().Result;
        ...
     }
}
```

But during startup I get the error:

_"Unable to resolve service for type 'Shiny.AndroidPlatform' while attempting to activate 'Shiny.Push.PushManager'._

I'm sure I've forgotten something fundamental here. But what?

**Comment by aritchie** [comment]
You don't directly inject shiny modules anywhere in your app.  There is extensions methods for all modules.  Please read the docs on this: https://shinylib.net/client/push/providers/   There is also a sample at: https://github.com/shinyorg/pushtester/tree/main/PushEndToEnd

> var result = PushManager?.RequestAccess().Result;

You will have tons of problems if you do stuff like this.  I'm going to go into the reasons things can deadlock here.

**Reply by konrad79** [reply]
But how do I get the RegistrationToken, that I need e.g. to send a (test) message from Firebase to exactly this device?

Using only
`builder.Services.AddPushFirebaseMessaging<MyPushDelegate>(...)`
doesn't produce this token.

And in the PushEndToEnd-Example a button must be clicked to run the code:

```
if (pushManager.RegistrationToken == null)
{
    var result = await pushManager.RequestAccess();
    if (result.Status != AccessState.Available)
    {
        await this.Dialogs.Alert("Failed to register - " + result.Status);
    }
    else
    {
        this.RegistrationToken = result.RegistrationToken;
    }  
}
```

where exactly the mentioned

`var result = await pushManager.RequestAccess();`

is used.

**Reply by aritchie** [reply]
> builder.Services.AddPushFirebaseMessaging<MyPushDelegate>(...)

This is the registration portion, so that you can dependency inject IPushManager

> var result = await pushManager.RequestAccess();

Wherever you want!?  It's a permission request, so it needs a UI screen of some sort if the request has to be made

**Reply by konrad79** [reply]
Yes, I want to call
`var result = await pushManager.RequestAccess();`
via the user clicking a button.

But I'm struggling to get the pushManager as DI into the code-behind. 

In the PushEndToEnd-Example the `pushManager `is injected in the `MainViewModel `- but there is no corresponding `.AddSingleton` or something like that in the `MauiProgram `class. Is that some magic in the deep of Shiny?

**Reply by aritchie** [reply]
The MainViewModel in my sample is using Prism and therefore participates in dependency injection.  How your pages/viewmodels play in dependency injection is your code and outside of the scope of Shiny.
